tools/testrunner/testrunner.go

```diff
@@ -0,0 +1,348 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package testrunner
+
+import (
+	"context"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"iter"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/jstemmer/go-junit-report/v2/junit"
+)
+
+type node struct {
+	children map[string]node
+}
+
+func (n node) visitor(path ...string) func(yield func(string, node) bool) {
+	return func(yield func(string, node) bool) {
+		for name, child := range n.children {
+			path := append(path, name)
+			if !yield(strings.Join(path, "/"), child) {
+				return
+			}
+			child.visitor(path...)(yield)
+		}
+	}
+}
+
+func (n node) walk() iter.Seq2[string, node] {
+	return n.visitor()
+}
+
+type attempt struct {
+	runner       *runner
+	number       int
+	junitXmlPath string
+	exitErr      *exec.ExitError
+	suites       junit.Testsuites
+}
+
+func (a *attempt) run(ctx context.Context, args []string) error {
+	args = append([]string{"--junitfile", a.junitXmlPath}, args...)
+	log.Printf("starting test attempt %d with args: %v", a.number, args)
+	cmd := exec.CommandContext(ctx, a.runner.gotestsumExecutable, args...)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	cmd.Stdin = os.Stdin
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *attempt) recordResult(exitErr *exec.ExitError) error {
+	var suites junit.Testsuites
+
+	f, err := os.Open(a.junitXmlPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	decoder := xml.NewDecoder(f)
+	if err := decoder.Decode(&suites); err != nil {
+		return err
+	}
+	a.exitErr = exitErr
+	a.suites = suites
+	return nil
+}
+
+func (a attempt) failures() []string {
+	root := node{children: make(map[string]node)}
+
+	for _, suite := range a.suites.Suites {
+		if suite.Failures == 0 {
+			continue
+		}
+		for _, tc := range suite.Testcases {
+			if tc.Failure == nil {
+				continue
+			}
+			n := root
```

Comment:
it took some time to realize that 'n' is something like 'node'.
I would call it treeNode (but I in general not a big fun of short names besides "this" substitutes)

Reply:
go rule: the less is the scope the shorter the name.